merge.py

```diff
@@ -1,98 +1,320 @@
+        try:
+            self.clone_configure(tmp_dir, pr)
+            pr_name = 'finish-pr-{}'.format(pr.get_num())
+            self.checkout(pr_name, tmp_dir, pr)
+            branch = '{}/{}'.format('apache', self.config['merge_branch'])
+            self.rebase(branch, tmp_dir, pr)
+            self.target_checkout(branch, tmp_dir, pr)
+            self.do_local_merge(pr_name, tmp_dir, pr)
+            self.verify(tmp_dir, pr)
+            self.push(tmp_dir, pr)
+            _clean_up(tmp_dir)
+            _print_flush('Merged successfully!')
+        except AssertionError as err:
+            _print_flush(err)
+        except EnvironmentError as err:
+            _print_flush(err)
+            return False
+        return True
+
+    def clone_configure(self, tmp_dir, pr):
+        """Clone repository and configure.
+
+        Args:
+            tmp_dir: Directory in which commands should be run.
+            pr: Pull request to merge.
+        Raises:
+            AssertionError: If command was not successful.
+            EnvironmentError: If PR comment couldn't be posted to Github.
+        """
+        _print_flush('Cloning repository and configuring.')
+        repo = self.config['repository']
+        repo_url = GITHUB_REPO_URL.format(org='apache', repo=repo)
+        clone_args = ['git', 'clone', '-b', self.config['merge_branch'],
+                      repo_url, tmp_dir]
+        try:
+            check_call(clone_args, cwd=tmp_dir)
+        except:
+            pr.post_error(
+                'Couldn\'t clone from github/{}/{}. Please try again.'.format(
+                    'apache', repo))
+            raise AssertionError('Clone Failed.')
+
+        try:
+            check_call(['git', 'remote', 'add', 'apache',
+                        APACHE_GIT.format(repo=repo)], cwd=tmp_dir)
+            check_call(['git', 'remote', 'rename', 'origin', 'github'],
+                       cwd=tmp_dir)
+            check_call('git config --local --add remote.github.fetch'
+                       ' "+refs/pull/*/head:refs/remotes/github/pr/*"',
+                       shell=True, cwd=tmp_dir)
+            check_call(['git', 'fetch', '--all'], cwd=tmp_dir)
+        except:
+            pr.post_error('Problem configuring repository. Please try again.')
+            raise AssertionError('Failed to set up.')
+        _print_flush('Initial configuration complete.')
+
+    def checkout(self, pr_name, tmp_dir, pr):
+        """Check out PR from Github.
+
+        Args:
+            pr_name: Name of pull request branch.
+            tmp_dir: Directory in which commands should be run.
+            pr: Pull request to merge.
+        Raises:
+            AssertionError: If command was not successful.
+            EnvironmentError: If PR comment couldn't be posted to Github.
+        """
+        _print_flush('Checking out PR.')
+        pr_num = pr.get_num()
+        ic_args = ['git', 'checkout', '-b', pr_name, 'github/pr/{}'.format(
```

Comment:
As a general high-level comment, one thing that might make these commands easier to read is if you always used a string to generate them (rather than a list) and you had an environment of data that was available to all of them (eg., the tmp_dir, the PR, etc.

Then this could be:

```
self.exec_git('git checkout -b {pr_name} github/pr/{pr_num}')
```

etc. using the standardly available environment. I think this would be easier to read because each command is easily consummable. You could even take a list of strings to execute:
```
self.exec_git([
  'git remote add apache {apache_repo}',
  'git remote rename origin github',
  'git config --local --add remote.github.fetch "+refs/pull/*/head/:refs/remotes/github/pr/*"',
  'git fetch --all'])
```

And that kind of helper makes it easy to write each of the methods you're writing here (which is essentially execute a series of commands) and it also makes it easy to see the sequence of commands being executed.